### PR TITLE
Towards DeepLabCut 3.0: SuperAnimal Inference

### PR DESCRIPTION
**Pull request by MMathisLab**
In 2024 we are moving to PyTorch (fully)🔥 ! This has been on the [roadmap for some years](https://github.com/DeepLabCut/DeepLabCut/blob/main/docs/roadmap.md), and in 2023 the team worked to completely overall the core of deeplabcut [with the generous support from CZI EOSS Program](https://chanzuckerberg.com/eoss/proposals/deeplabcut-a-software-package-for-animal-pose-estimation/). It will feel (mostly) seamless to users to move to this version, and it will have the same high-level API. 

This PR introduces (only) the model inference (`video_analysis`, etc.) for **DLC3.0 pytorch trained models**. 
- The first models that we release are our new foundation animal pose models, which are documented in [Ye et al. on arXiv](https://arxiv.org/abs/2203.07436). 
- The Model Cards and details can be found [here](https://huggingface.co/mwmathis/DeepLabCutModelZoo-SuperAnimal-Quadruped) and [here](https://huggingface.co/mwmathis/DeepLabCutModelZoo-SuperAnimal-TopViewMouse). 

-There are a huge number of people behind this, so check out the contributors who have really worked incredibly hard to build a well engineered and tested code base!
- Lastly, better API documentation is forthcoming 🥳, but you can check out the code in the meantime.

### CODE REVIEW

File: deeplabcut/modelzoo/utils.py [hunk 1]
```diff
@@ -11,11 +11,56 @@
...
+from deeplabcut.utils.auxiliaryfunctions import get_deeplabcut_path
 
-    dlc_path = deeplabcut.utils.auxiliaryfunctions.get_deeplabcut_path()
-    json_path = os.path.join(dlc_path, "modelzoo", "models.json")
-    with open(json_path) as file:
-        return json.load(file)
+
+def parse_project_model_name(superanimal_name: str) -> str:
+    """
+    TODO
```
Comment (MMathisLab) [inline deeplabcut/modelzoo/utils.py]: @n-poulsen lets add some quick docstrings here.

File: deeplabcut/modelzoo/webapp/inference.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
+from dataclasses import dataclass
```
Comment (MMathisLab) [inline deeplabcut/modelzoo/webapp/inference.py:11]: add header @stes

File: deeplabcut/pose_estimation_pytorch/__init__.py [hunk 3]
```diff
@@ -0,0 +1,12 @@
+from deeplabcut.pose_estimation_pytorch.apis import (
```
Comment (MMathisLab) [inline deeplabcut/pose_estimation_pytorch/__init__.py:11]: add header @n-poulsen @stes -- adjust the make header file

File: deeplabcut/pose_estimation_pytorch/config/tokenpose/tokenpose_base.yaml [hunk 4]
```diff
@@ -0,0 +1,48 @@
+# TODO: MATCH https://github.com/leeyegy/TokenPose/blob/main/experiments/coco/tokenpose/tokenpose_b_256_192_patch43_dim192_depth12_heads8.yaml
```
Comment (MMathisLab) [inline deeplabcut/pose_estimation_pytorch/config/tokenpose/tokenpose_base.yaml]: @n-poulsen unclear TODO

File: deeplabcut/pose_estimation_pytorch/config/utils.py [hunk 5]
```diff
@@ -0,0 +1,188 @@
+"""Util functions to create pytorch pose configuration files"""
```
Comment (MMathisLab) [inline deeplabcut/pose_estimation_pytorch/config/utils.py:11]: add header @stes

File: deeplabcut/pose_estimation_pytorch/data/dlcloader.py [hunk 6]
```diff
@@ -0,0 +1,182 @@
+from __future__ import annotations
```
Comment (MMathisLab) [inline deeplabcut/pose_estimation_pytorch/data/dlcloader.py:11]: add header @n-poulsen

File: deeplabcut/pose_estimation_pytorch/data/helper.py [hunk 7]
```diff
@@ -0,0 +1,86 @@
+from __future__ import annotations
```
Comment (MMathisLab) [inline deeplabcut/pose_estimation_pytorch/data/helper.py:11]: add header @n-poulsen

File: deeplabcut/pose_estimation_pytorch/data/postprocessor.py [hunk 8]
```diff
@@ -0,0 +1,434 @@
+"""Post-process predictions made by models"""
```
Comment (MMathisLab) [inline deeplabcut/pose_estimation_pytorch/data/postprocessor.py:11]: add header @n-poulsen

File: deeplabcut/pose_estimation_pytorch/data/preprocessor.py [hunk 9]
```diff
@@ -0,0 +1,328 @@
+"""Helpers to run preprocess data before running inference"""
```
Comment (MMathisLab) [inline deeplabcut/pose_estimation_pytorch/data/preprocessor.py:11]: add header @n-poulsen

File: deeplabcut/pose_estimation_pytorch/data/transforms.py [hunk 10]
```diff
@@ -0,0 +1,330 @@
+from __future__ import annotations
```
Comment (MMathisLab) [inline deeplabcut/pose_estimation_pytorch/data/transforms.py:11]: add header @n-poulsen

File: deeplabcut/pose_estimation_pytorch/models/__init__.py [hunk 11]
```diff
@@ -0,0 +1,10 @@
+from deeplabcut.pose_estimation_pytorch.models.backbones.base import BACKBONES
```
Comment (MMathisLab) [inline deeplabcut/pose_estimation_pytorch/models/__init__.py:11]: add header @n-poulsen

File: deeplabcut/pose_estimation_pytorch/models/criterions/__init__.py [hunk 12]
```diff
@@ -0,0 +1,14 @@
+from deeplabcut.pose_estimation_pytorch.models.criterions.aggregators import (
```
Comment (MMathisLab) [inline deeplabcut/pose_estimation_pytorch/models/criterions/__init__.py:11]: add header

File: deeplabcut/pose_estimation_pytorch/models/criterions/utils.py [hunk 13]
```diff
@@ -0,0 +1,40 @@
+from __future__ import annotations
```
Comment (MMathisLab) [inline deeplabcut/pose_estimation_pytorch/models/criterions/utils.py:11]: add header

File: deeplabcut/pose_estimation_pytorch/models/detectors/__init__.py [hunk 14]
```diff
@@ -0,0 +1,5 @@
+from deeplabcut.pose_estimation_pytorch.models.detectors.base import (
```
Comment (MMathisLab) [inline deeplabcut/pose_estimation_pytorch/models/detectors/__init__.py:11]: add header

File: deeplabcut/pose_estimation_pytorch/models/heads/__init__.py [hunk 15]
```diff
@@ -0,0 +1,5 @@
+from deeplabcut.pose_estimation_pytorch.models.heads.base import HEADS, BaseHead
```
Comment (MMathisLab) [inline deeplabcut/pose_estimation_pytorch/models/heads/__init__.py:11]: add header

File: deeplabcut/pose_estimation_pytorch/models/modules/__init__.py [hunk 16]
```diff
@@ -0,0 +1,8 @@
+from deeplabcut.pose_estimation_pytorch.models.modules.conv_block import (
```
Comment (MMathisLab) [inline deeplabcut/pose_estimation_pytorch/models/modules/__init__.py:11]: add header

File: pyproject.toml [hunk 17]
```diff
@@ -0,0 +1,20 @@
+[tool.yapf]
```
Comment (MMathisLab) [inline pyproject.toml:1]: what is this file for? @n-poulsen

File: setup.py [hunk 18]
```diff
@@ -1,20 +1,35 @@
...
+        for p in subdir.iterdir():
+            if p.suffix == ".yaml":
+                pytorch_configs.append(str(p))
+
+    return pytorch_configs
+
+
 setuptools.setup(
     name="deeplabcut",
     version="2.3.8",
```
Comment (MMathisLab) [inline setup.py:35]: lets bump the version to 3.0.0alpha cc @AlexEMG @n-poulsen @jeylau

File: setup.py [hunk 19]
```diff
@@ -37,10 +54,12 @@
         "pandas>=1.0.1,!=1.5.0",
         "scikit-image>=0.17",
         "scikit-learn>=1.0",
-        "scipy>=1.9",
+        "scipy>=1.4,<1.11.0",
         "statsmodels>=0.11",
         "tables>=3.7.0",
-        "torch<=1.12",
+        "timm",
+        "torch>=2.0.0",
```
Comment (MMathisLab) [inline setup.py:61]: is the older transformer for re-ID working with torch > 2? cc @yeshaokai please test and confirm.

File: deeplabcut/pose_estimation_pytorch/post_processing/identity.py [hunk 20]
```diff
@@ -0,0 +1,36 @@
+"""Functions to assign identity to predictions from an identity head"""
```
Comment (MMathisLab) [inline deeplabcut/pose_estimation_pytorch/post_processing/identity.py:11]: add header

File: deeplabcut/pose_estimation_pytorch/post_processing/__init__.py [hunk 21]
```diff
@@ -0,0 +1,4 @@
+from deeplabcut.pose_estimation_pytorch.post_processing.match_predictions_to_gt import (
```
Comment (MMathisLab) [inline deeplabcut/pose_estimation_pytorch/post_processing/__init__.py:11]: add header

File: deeplabcut/pose_estimation_pytorch/modelzoo/utils.py [hunk 22]
```diff
@@ -0,0 +1,105 @@
+import inspect
```
Comment (MMathisLab) [inline deeplabcut/pose_estimation_pytorch/modelzoo/utils.py:11]: add header

**Review by MMathisLab — COMMENTED** [review]
- @jeylau @n-poulsen @nastya236 @stes  please run the automated header code; some missing as noted.
- minimal useage docs would be ideal for users to test asap 
- please check docstrings

Otherwise awesome :)

**Comment by MMathisLab** [comment]
@n-poulsen please check codespell; i.e, exclusion of some files is needed, and two clear fixes needed :D :

```bash 
Error: ./deeplabcut/pose_estimation_pytorch/config/make_pose_config.py:108: BU ==> BY, BE, BUT, BUG, BUN, BUD, BUY, BUM
Error: ./deeplabcut/pose_estimation_pytorch/config/make_pose_config.py:210: bu ==> by, be, but, bug, bun, bud, buy, bum
Error: ./deeplabcut/pose_estimation_pytorch/config/base/base.yaml:34: bu ==> by, be, but, bug, bun, bud, buy, bum
Error: ./deeplabcut/pose_estimation_pytorch/apis/utils.py:470: bu ==> by, be, but, bug, bun, bud, buy, bum
Error: ./deeplabcut/pose_estimation_pytorch/apis/analyze_videos.py:235: BU ==> BY, BE, BUT, BUG, BUN, BUD, BUY, BUM
Error: ./deeplabcut/pose_estimation_pytorch/modelzoo/inference.py:108: BU ==> BY, BE, BUT, BUG, BUN, BUD, BUY, BUM
Error: ./deeplabcut/pose_estimation_pytorch/data/postprocessor.py:289: BU ==> BY, BE, BUT, BUG, BUN, BUD, BUY, BUM
Error: ./deeplabcut/pose_estimation_pytorch/data/transforms.py:[17](https://github.com/DeepLabCut/DeepLabCut/actions/runs/7360532599/job/20036759485?pr=2472#step:4:18)3: overlayed ==> overlaid
Error: ./deeplabcut/pose_estimation_pytorch/runners/base.py:28: BU ==> BY, BE, BUT, BUG, BUN, BUD, BUY, BUM
Error: ./deeplabcut/pose_estimation_pytorch/runners/schedulers.py:34: beggining ==> beginning
```

**Comment by MMathisLab** [comment]
@nastya236 and @n-poulsen tests are failing due to missing import (I think), so please double check, thanks!

```bash
==================================== ERRORS ====================================
______________ ERROR collecting tests/test_predict_supermodel.py _______________
ImportError while importing test module '/home/runner/work/DeepLabCut/DeepLabCut/tests/test_predict_supermodel.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/opt/hostedtoolcache/Python/3.9.18/x64/lib/python3.9/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
tests/test_predict_supermodel.py:13: in <module>
    from deeplabcut.modelzoo.api import superanimal_inference
E   ModuleNotFoundError: No module named 'deeplabcut.modelzoo.api'
```

**Comment by nastya236** [comment]
I forgot to update tensorflow test file for superanimal inference. @MMathisLab  now is fixed here: [PR#2473](https://github.com/DeepLabCut/DeepLabCut/pull/2473).

**Comment by MMathisLab** [comment]
A few more tests need checked like for dekr: 

FAILED tests/test_dekr_predictor.py::test_DEKRPredictor - Failed: DEKRPredictor forward pass raised an exception: tuple indices must be integers or slices, not str

File: pyproject.toml (hunk 17 above)
Comment (n-poulsen) [reply pyproject.toml:1]: It's a [configuration file](https://packaging.python.org/en/latest/guides/writing-pyproject-toml/) used by the linters we use (e.g. [Black](https://black.readthedocs.io/en/stable/usage_and_configuration/the_basics.html#configuration-via-a-file), isort) so we all have that we all have a common configuration

It was added by @jeylau, and I think we should actually update it now - it configures YAPF styling but we use Black.

**Comment by n-poulsen** [comment]
I've fixed the file headers, but only for files in the `deeplabcut/modelzoo` and `deeplabcut/pose_estimation_pytorch` folders. I believe most headers need to be updated (switching from `master` to `main` as the default branch), but this should be done in a separate PR. The following was also done:

 [7425cc8](https://github.com/DeepLabCut/DeepLabCut/pull/2472/commits/7425cc8cc8b0090cad089e8029ce672c62237b7d)
- fixed typos
- removed duplicate file headers
- updated the codespell configuration to ignore `bu`, `BU`, `td`, `TD`

[9d5e2dd](https://github.com/DeepLabCut/DeepLabCut/pull/2472/commits/9d5e2dddf7379f55c032ec03e4e3c113cc3c9efa)
- added a docstring for `parse_project_model_name` in `deeplabcut/modelzoo/utils.py`
- removed the failing test `tests/test_dekr_predictor.py` (a legacy test that I didn't catch when preparing the branch)
- improved the description of the TODO for the TokenPose base configuration

Some checks still seems to be failing, but that's due to an [incorrect hash](https://github.com/DeepLabCut/DeepLabCut/actions/runs/7385072804/job/20089134284?pr=2472) when installing `FFmpeg` on OSX, which I don't believe is an issue from our end.

**Comment by MMathisLab** [comment]
This PR/branch will stay for historical purposes but is going to be superseded by a new, fuller integration, PR ASAP.

**Pull request closed without merging**